hydroflow/src/builder/hydroflow_builder.rs

```diff
@@ -0,0 +1,124 @@
+    /// using the Surface API.
+    pub fn make_handoff<H, T>(
+        &mut self,
+    ) -> (HandoffPushSurfaceReversed<H, T>, HandoffPullSurface<H>)
+    where
+        H: Handoff + CanReceive<T>,
+    {
+        let push_port = Default::default();
+        let push = HandoffPushSurfaceReversed::new(Rc::clone(&push_port));
+
+        let pull_port = Default::default();
+        let pull = HandoffPullSurface::new(Rc::clone(&pull_port));
+
+        self.handoff_connectors.push(Box::new(move |hydroflow| {
+            if let (Some(output_port), Some(input_port)) = (push_port.take(), pull_port.take()) {
+                hydroflow.add_edge(output_port, input_port);
+            } else {
+                panic!("Handoff was never connected!!"); // TODO(mingwei): more informative error messages.
+            }
+        }));
+        (push, pull)
+    }
+
+    /// Adds a `pivot` created via the Surface API.
+    pub fn add_subgraph<Pull, Push>(&mut self, pivot: PivotSurface<Pull, Push>)
+    where
+        Pull: 'static + PullSurface,
+        Push: 'static + PushSurfaceReversed<ItemIn = Pull::ItemOut>,
+    {
+        let ((pull_connect, push_connect), (mut pull_build, mut push_build)) = pivot.into_parts();
+
+        let (input_ports, output_ports) = self
+            .hydroflow
+            .add_subgraph::<_, Pull::InputHandoffs, Push::OutputHandoffs>(
+                move |_ctx, recv_ctx, send_ctx| {
+                    let pull = pull_build.build(recv_ctx);
+                    let push = push_build.build(send_ctx);
+                    let pivot = Pivot::new(pull, push);
+                    pivot.run();
+                },
+            );
+
+        pull_connect.connect(input_ports);
+        push_connect.connect(output_ports);
+    }
+
+    /// Creates a new external channel input.
+    pub fn add_channel_input<T, W>(&mut self) -> (Input<T, SyncSender<T>>, HandoffPullSurface<W>)
+    where
+        T: 'static,
+        W: 'static + Handoff + CanReceive<T>,
+    {
+        let (input, output_port) = self.hydroflow.add_channel_input();
+
+        let pull_port = Default::default();
+        let pull = HandoffPullSurface::new(Rc::clone(&pull_port));
+
+        self.handoff_connectors.push(Box::new(move |hydroflow| {
+            if let Some(input_port) = pull_port.take() {
+                hydroflow.add_edge(output_port, input_port);
+            } else {
+                panic!("Channel input was never connected!!"); // TODO(mingwei): more informative error messages.
+            }
+        }));
+
+        (input, pull)
+    }
+
+    pub fn build(mut self) -> Hydroflow {
+        for handoff_connector in self.handoff_connectors {
+            // TODO(mingwei): be more principled with this.
+            (handoff_connector)(&mut self.hydroflow);
+        }
+        self.hydroflow
+    }
+
+    /// Start a new branch for teeing
+    ///
+    /// This is needed due to the ownership order:
```

Comment:
nit: the `:` makes it seems like this is about to lead into an explanation but none follows?